### DISCUSSION TITLE
Help Needed: Configuring YAML for handoff_forecast_lstm Model

### DISCUSSION THREAD

**Discussion by gigigi0712** [neuralhydrology/neuralhydrology#219] [Q&A]
I am attempting to replicate the experimental setup described in the paper "Global Prediction of Extreme Floods in Ungauged Watersheds" using the NeuralHydrology framework. To achieve this, I have prepared a YAML configuration file for running experiments with the handoff_forecast_lstm model in NeuralHydrology. However, I’m not sure if my configuration is correct and, in particular, I’m confused about how to set up the CMAL head and its specific parameters.

Below is the portion of my YAML configuration file:
```
# --- Experiment configurations --------------------------------------------------------------------

# experiment name, used as folder name
experiment_name: xxxxxxx

# place to store run directory (if empty runs are stored in $cwd$/runs/)
run_dir:

# files to specify training, validation and test basins (relative to code root or absolute path)
train_basin_file: 531_basin_list.txt
validation_basin_file: 531_basin_list.txt
test_basin_file: 531_basin_list.txt

# training, validation and test time periods (format = 'dd/mm/yyyy')
train_start_date: '01/10/1999'
train_end_date: '30/09/2008'
validation_start_date: '01/10/1980'
validation_end_date: '30/09/1989'
test_start_date: '01/10/1989'
test_end_date: '30/09/1999'

# fixed seed, leave empty to use a random seed
seed: 123

# which GPU (id) to use [in format of cuda:0, cuda:1 etc, or cpu or None]
device: cuda:0

# --- Validation configuration ---------------------------------------------------------------------

# specify after how many epochs to perform validation
validate_every: 5

# specify how many random basins to use for validation
validate_n_random_basins: 531

# specify which metrics to calculate during validation (see codebase.evaluation.metrics)
metrics:
- NSE
- RMSE
- KGE

# --- Model configuration --------------------------------------------------------------------------

# base model type [lstm, ealstm, cudalstm, embcudalstm, shortcutlstm, dropoutlstm, cudalstminitialh, handoff_forecast_lstm ]
# (has to match the if statement in modelzoo/__init__.py)
model: handoff_forecast_lstm

# An integer number of timesteps where forecast data overlaps with hindcast data. This does not add to the forecast_sequence_length, and must be no larger than the forecast_sequence_length
forecast_overlap: 0
forecast_seq_length: 1
state_handoff_network: {"type": "full_model", "hiddens": 128}

# prediction head [regression, mdn, umal]. Define the head specific parameters below
head: cmal

# ----> Regression settings <----
output_activation: linear

# ----> General settings <----

# Number of cell states of the LSTM
hidden_size: 128

# Initial bias value of the forget gate
initial_forget_bias: 3

# Dropout applied to the output of the LSTM
output_dropout: 0.4

# --- Training configuration -----------------------------------------------------------------------

# specify optimizer [Adam, Adadelta]
optimizer: Adam

# specify loss [MSE, NSE, RMSE, UMALLoss, MDNLoss]
loss: CMALloss
n_distributions: 2
n_samples: 500
negative_sample_handling: 'clip'
negative_sample_max_retries: 5
mc_dropout: False

# specify learning rates to use starting at specific epochs (0 is the initial learning rate)
learning_rate:
  0: 1e-3
  1: 5e-4

# Mini-batch size
batch_size: 256

# Number of training epochs
epochs: 50

# If True, clips norm of gradients
clip_gradient_norm: 1

# Defines which time steps are used to calculate the loss. Can't be larger than seq_length
predict_last_n: 1

# Length of the input sequence
seq_length: 366

# Number of parallel workers used in the data pipeline
num_workers: 0

# Log the training loss every n steps
log_interval: 5

# If true, writes logging results into tensorboard file
log_tensorboard: True

# Save model weights every n epochs
save_weights_every: 1

# Store the results of the validation to disk
save_validation_results: True

# --- Data configurations --------------------------------------------------------------------------

dataset: camels_us

# Path to CAMELS data set
data_dir: ....../CAMELS_US

# Forcing product [daymet, maurer, maurer_extended, nldas, nldas_extended]
# can be either a list of forcings or a single forcing product
forcings:
- daymet

# variables to use as time series input (names match the data file column headers)
# Note: In case of multiple input forcing products, you have to append the forcing product behind
# each variable. E.g. 'prcp(mm/day)' of the daymet product is 'prcp(mm/day)_daymet'
dynamic_inputs:
- prcp(mm/day)
- srad(W/m2)
- tmax(C)
- tmin(C)
- vp(Pa)

# which columns to use as target
target_variables:
- QObs(mm/d)

forecast_inputs:
- prcp(mm/day)
- srad(W/m2)
- tmax(C)
- tmin(C)
- vp(Pa)

hindcast_inputs:
- prcp(mm/day)
- srad(W/m2)
- tmax(C)
- tmin(C)
- vp(Pa)

static_attributes:
- elev_mean
- slope_mean
- area_gages2
- frac_forest
- lai_max
- lai_diff
- gvf_max
- gvf_diff
- soil_depth_pelletier
- soil_depth_statsgo
- soil_porosity
- soil_conductivity
- max_water_content
- sand_frac
- silt_frac
- clay_frac
- carbonate_rocks_frac
- geol_permeability
- p_mean
- pet_mean
- aridity
- frac_snow
- high_prec_freq
- high_prec_dur
- low_prec_freq
- low_prec_dur
```

Specific Questions:

CMAL Head Parameters:

n_distributions: Is setting this to 3 appropriate for modeling the mixture distributions in this context?
n_samples: I have chosen 100 samples per time-step; does this align with the experimental setup in the paper?
negative_sample_handling: I opted for "truncate" to manage negative samples. Are there specific scenarios where "clip" or "none" would be more suitable?
mc_dropout: I have enabled Monte Carlo dropout during inference. Is this consistent with the methods used in the study?
Overall Configuration:

Does this YAML file accurately reflect the experimental setup described in the paper?
Are there any parameters or settings that I might have overlooked or should adjust to better replicate the study?
I have referred to the original paper (https://www.nature.com/articles/s41586-024-07145-1) for guidance. However, detailed configuration specifics were not fully clear to me. Any insights, suggestions, or shared experiences in replicating this study would be highly valuable.

I appreciate any detailed guidance or sample configurations from those who have experience with NeuralHydrology and the CMAL head.

**Comment by kratzert** [comment]
Hi @gigigi0712 

I think the best would be to reach out to Grey Nearing (the contact author on the paper and the person who implemented the model here) for details on this. I personally never trained any of the forecast models with NeuralHydrology so it would be hard for me to tell.

Best,
Freddy

**Comment by grey-nearing** [comment]
Hi [gigigi0712](https://github.com/gigigi0712),

Thank you for your question. There are a few things to know. 

First, it is not possible to replicate the results from the Nearing et al. [2024] paper with NeuralHydrology. The model used in that study is included in the NeuralHydrology codebase (it is the Handoff model that you are working with), but the data used for that study is not publicly available, due to licensing issues that are unfortunately out of the control of the authors. 

Please understand that one major difference between the data you are using in this config file vs. the data used in the 2024 study is that there is no actual meteorological forecast data in the dataset you are using, so you will not get an accurate assessment of forecast skill. We published an open source dataset that includes meteorological forecasts here: https://arxiv.org/abs/2411.09459. This is NOT the same dataset used in the 2024 paper, but is closer than anything we or other research groups were able to make publicly available before (e.g., Caravan, CAMELS, etc.).

One difference that I see in your config file vs. the experimental setup from that paper is the `forecast_seq_length` parameter, which should be set to a forecast time horizon of 7 days instead of 1 day. Similarly, the `predict_last_n` parameter we used was 7(instead of 1), meaning that we calculate the loss function during training on the full 7-day forecast sequence.

Regarding CMAL, using 3 distributions is great. Make sure that you do not use output dropout with the CMAL head (set output_dropout=0). MC Dropout was not used in the 2024 study, however I just pinged the author of the original NeuralHydrology CMAL paper [1] and he did use MC Droupout with CMAL for higher-order uncertainty estimation (see Fig 11 in that paper). The sampling parameter is used during inference, not during training, to extract percentiles from the predicted probability distributions (this is done by sampling instead of analytically).

[1] https://hess.copernicus.org/articles/26/1673/2022/#section3

**Reply by brett-snider** [reply]
@grey-nearing thanks for this answer! This answer has prompted a couple other follow up questions:

If my forecast_seq_length = 7 days and predict_last_n is 1, is it predicting the flow for the 7th day? or first day of forecast? 

If i want the forecast_overlap to be 7 days and then to predict 7 days into the future do i set up forecast_seq_length = 14, and forecast_overlap=7 and predict_last_n=7 to get predictions for Day 1:7?
